---
title: "Selection of model parameters for the gene x environment studies on the Gulf War Illness"
author: "Lucas Zhang (Supervised by Elizabeth Hauser, PhD)"
date: "2022 Spring"
output: html_document
runtime: shiny
---

## 1. Introduction

Gulf War illness (GWI) as an enduring multi-symptom syndrome affects more than one-third of the U.S. veterans who served during the Gulf War from 1990 to 1991 (White et al., 2016). The illness is associated with a variety of acute and/or chronic symptoms including fatigue, widespread pain, cognitive difficulties, skin rashes and gastrointestinal and respiratory disorders (White et al., 2016). Considerable resources have been devoted to characterizing the illness, investigating its possible causes and exploring the underlying mechanisms. While the detailed molecular and cellular mechanisms remain uncharacterized, possible causes have been discussed including psychological trauma and exposures to chemicals such as chemical warfare agents, pyridostigmine bromide, pesticides, oil well fires and depleted uranium.

To determine the molecular mechanisms of GWI and improve the well-being of veterans, the US Department of Veteran Affairs Cooperative Studies Program launched a study of the Gulf War Era Cohort and Biorepository (GWECB). The GWECB cohort comprises 1343 subjects. The veteransâ€™ medical records, blood samples and survey data including their demographic information, health behavior and environmental exposures have been collected (Gifford et al., 2021).  This dataset allows genome-wide association studies which aim to determine the penetrance function of GWI and estimate the gene-environmental interaction effect (Radhakrishnan et al., 2021). Specifically, a penetrance function relates the proportion of veterans who have GWI to their different genotypes and environmental exposure histories. 

## 2. Methods and Materials

## 3. Results and discussion

## How will different penetrance functions influece the GWI prevalance?

The GWI have a very high prevanlance estimated at ~ 15%-80%. Here we want to explore the parameter space of different parameters in the genetic model, and therefore, to identify possible combinations of allele frequency and penetrance functions. As for the following genetic model for a specific biallelic susceptibility locus of 
$$K = p^2f_{AA}+2p(1-p)f_{Aa}+(1-p)^2f_{aa}$$,
where $p$ indicates the allele frequency of *A*, $p^2$, $2p(1-p)$ and $(1-p)^2$ are frequencies of *AA*, *Aa* and  *aa* genotypes, and $f_{AA}$, $f_{Aa}$ and $f_{aa}$ are their penetrance functions, respectively.

```{r initialization, message=FALSE,warning=FALSE}
library(tidyverse)
library(knitr)
library(kableExtra)
library(plotly)
```

```{r interative figure, echo = FALSE}
p = seq(0, 1, 0.01)
shiny::sliderInput("f_GG", label = "f_GG",
                   min = 0, max = 1, value = 1,
                   step = 0.01)
shiny::sliderInput("f_AG", label = "f_AG",
                   min = 0, max = 1, value = 0.5,
                   step = 0.01)
shiny::sliderInput("f_AA", label = "f_AA",
                   min = 0, max = 1, value = 0,
                   step = 0.01)
plotly::renderPlotly({
  K = p^2*(input$f_GG) + 2*p*(1-p)*(input$f_AG) +
            (1-p)^2*(input$f_AA)
  ggplotly(qplot(x= p, y=K) + theme_bw())
})
```


## How will gene-environment interactions influence the GWI prevelence? 
Considering the following genetic model for Gene-Environment interactions:

$$logit(Y)=\beta_0+\beta_1 \cdot SNP + \beta_2 \cdot E + \beta_3 \cdot SNP \cdot E$$,

where $SNP$ is the (additive) genetic model, ranging from 0-2. 0: the AA genotype, 1: GA genotype, 2: GG genotype. $E$ is the environmental exposure, ranging from 0-3. 0: No exposure, 1: less than 7 days, 2: 7-30 days, 3: more than 30 days.

Observations:

| Genotype      | Exposure      | Freq  |
| ------------- |:-------------:| -----:|
| AA =  0       | 0  | 0.20 |
| AA =  0       | 1  | 0.23 |
| AA =  0       | 2  | 0.25 |
| AA =  0       | 3  | 0.23 |
| AG =  1       | 0  | 0.20 |
| AG =  1       | 1  | 0.26 |
| AG =  1       | 2  | 0.26 |
| AG =  1       | 3  | 0.40 |
| GG =  2       | 0  | 0.20 |
| GG =  2       | 1  | 0.40 |
| GG =  2       | 2  | 0.35 |
| GG =  2       | 3  | 0.46 |

The best fit estimates:
```{r observations, message=FALSE}
G = rep(0:2, each = 4)
E = rep(0:3, times= 3)
Freq = c(0.20, 0.23, 0.25, 0.23, 
         0.20, 0.26, 0.26, 0.40,
         0.20, 0.40, 0.35, 0.46)
OBS <- data.frame(G = factor(G), E = E, Freq = Freq, 
                  logit_Freq = log(Freq/(1-Freq)))
OBS.model <- lm(logit_Freq ~ G + E + G * E, data = OBS)
kbl(OBS.model$coefficients, 
    caption = "Best Fit Coefficients", 
    col.names = c("Coefficients"), 
    position = "h", digits = 4) %>%
  kable_classic(full_width = F)
```

```{r interative figure2, echo=FALSE, fig.width=8, fig.height=6}
coef_i<-OBS.model$coefficients
shiny::sliderInput("beta0", label = "beta0",
                   min = -3, max = 0, value = coef_i[1],
                   step = 0.01)
shiny::sliderInput("beta1_1", label = "beta1_1",
                   min = -0.3, max = 0.3, value = coef_i[2],
                   step = 0.002)
shiny::sliderInput("beta1_2", label = "beta1_2",
                   min = -0.3, max = 0.3, value = coef_i[3],
                   step = 0.002)
shiny::sliderInput("beta2", label = "beta2",
                   min = 0, max = 0.3, value = coef_i[4],
                   step = 0.001)
shiny::sliderInput("beta3_1", label = "beta3_1",
                   min = 0, max = 0.5, value = coef_i[5],
                   step = 0.002)
shiny::sliderInput("beta3_2", label = "beta3_2",
                   min = 0, max = 0.5, value = coef_i[6],
                   step = 0.002)
plotly::renderPlotly({
  Geno_risk <- 
    data.frame(Genotype = factor(c("AA", "AG", "GG")), 
               risk = c(0, input$beta1_1, input$beta1_2))
  p <- ggplot(Geno_risk, aes(Genotype, risk)) + geom_point() + theme_bw()
  ggplotly(p)
})
plotly::renderPlotly({
  Geno_risk <- 
    data.frame(Genotype = factor(c("AA", "AG", "GG")), 
               risk = c(0, input$beta1_1, input$beta1_2))
  p <- ggplot(Geno_risk, aes(Genotype, risk)) + geom_point()
  ggplotly(p)
  
  OBS.model.new <- OBS.model
  OBS.model.new$coefficients <- c(input$beta0, input$beta1_1, input$beta1_2,
                                  input$beta2, input$beta3_1, input$beta3_2)
  
  logit_Freq_sim = predict(obj = OBS.model.new, type = "response", newdata = OBS)
  Freq_sim = exp(logit_Freq_sim) / (1+exp(logit_Freq_sim))
  sim <- data.frame(Genotype = factor(G),
                    Exposure = factor(E),
                    Yes = Freq_sim, No = 1- Freq_sim)
  levels(sim$Genotype) = c("AA","GA","GG")
  levels(sim$Exposure) = c("None", "1-6 days", 
                           "7-30 days", "31+ days")
  sim2 <- reshape2::melt(sim, id.vars = c(1,2), value.name="Frequency")
  sim2$variable <- factor(sim2$variable, levels = c("No","Yes"))
  p <- ggplot(sim2, aes(x = Exposure, y = Frequency, color = variable, fill = variable)) + 
    geom_bar(stat="identity")+facet_grid(.~Genotype) + theme_bw()
  ggplotly(p)
})
```

# Simulation studies assuming unequal participation rate among two **phenotypes**

```{r simulations, echo=FALSE, fig.width=8, fig.height=6}
shiny::sliderInput("beta0_new", label = "beta0",
                   min = -3, max = 0, value = coef_i[1],
                   step = 0.01)
shiny::sliderInput("alpha1", 
                   label = "Participation Rate of GWI Population",
                   min = 0, max = 1, value = 0.15,
                   step = 0.01)
shiny::sliderInput("alpha2", 
                   label = "Participation Rate of Healthy Population",
                   min = 0, max = 1, value = 0.10,
                   step = 0.01)

plotly::renderPlotly({
  Geno_risk <- 
    data.frame(Genotype = factor(c("AA", "AG", "GG")), 
               risk = c(0, input$beta1_1, input$beta1_2))
  OBS.model.new <- OBS.model
  OBS.model.new$coefficients <- c(input$beta0_new, input$beta1_1, input$beta1_2,
                                  input$beta2, input$beta3_1, input$beta3_2)
  
  logit_Freq_sim = predict(obj = OBS.model.new, type = "response", newdata = OBS)
  Freq_sim = exp(logit_Freq_sim) / (1+exp(logit_Freq_sim))
  sim <- data.frame(Genotype = factor(G),
                    Exposure = factor(E),
                    Yes = Freq_sim, 
                    No = (1- Freq_sim) )
  sim$Yes <- sim$Yes * input$alpha1
  sim$No <- sim$No * input$alpha2
  sim$total <- sim$Yes + sim$No
  sim$Yes <- sim$Yes / sim$total
  sim$No <- sim$No / sim$total
  sim <- sim[, 1:4]
  levels(sim$Genotype) = c("AA","GA","GG")
  levels(sim$Exposure) = c("None", "1-6 days", 
                           "7-30 days", "31+ days")
  sim2 <- reshape2::melt(sim, id.vars = c(1,2), value.name="Frequency")
  sim2$variable <- factor(sim2$variable, levels = c("No","Yes"))
  p <- ggplot(sim2, aes(x = Exposure, y = Frequency, color = variable, fill = variable)) + 
    geom_bar(stat="identity")+facet_grid(.~Genotype) + theme_bw()
  ggplotly(p)
})

```

# Simulation studies assuming missclassification of GWI phenotype

```{r simulations2, echo=FALSE, fig.width=8, fig.height=6}
shiny::sliderInput("beta0_new2", label = "beta0",
                   min = -3, max = 0, value = coef_i[1],
                   step = 0.01)
shiny::sliderInput("sensitivity", 
                   label = "True Positive Rate (% Diagonosed GWI among all GWI Patients)",
                   min = 0, max = 100, value = 100,
                   step = 1)
shiny::sliderInput("specificity", 
                   label = "True Negative Rate (% Considered GWI-free among all healthy population)",
                   min = 0, max = 100, value = 100,
                   step = 1)

plotly::renderPlotly({
  Geno_risk <- 
    data.frame(Genotype = factor(c("AA", "AG", "GG")), 
               risk = c(0, input$beta1_1, input$beta1_2))
  OBS.model.new <- OBS.model
  OBS.model.new$coefficients <- c(input$beta0_new2, input$beta1_1, input$beta1_2,
                                  input$beta2, input$beta3_1, input$beta3_2)
  
  logit_Freq_sim = predict(obj = OBS.model.new, type = "response", newdata = OBS)
  Freq_sim = exp(logit_Freq_sim) / (1+exp(logit_Freq_sim))
  sim <- data.frame(Genotype = factor(G),
                    Exposure = factor(E),
                    Yes = Freq_sim, 
                    No = (1- Freq_sim) )
  sim$Diagnosed_Yes <- sim$Yes * input$sensitivity/100 + sim$No * (1-input$specificity/100)
  sim$Diagnosed_No <- sim$Yes * (1-input$sensitivity/100) + sim$No * input$specificity/100
  sim$total <- sim$Diagnosed_Yes + sim$Diagnosed_No
  sim$Diagnosed_Yes <- sim$Diagnosed_Yes / sim$total
  sim$Diagnosed_No <- sim$Diagnosed_No / sim$total
  sim <- sim[, c("Genotype", "Exposure", "Diagnosed_Yes", "Diagnosed_No")]
  levels(sim$Genotype) = c("AA","GA","GG")
  levels(sim$Exposure) = c("None", "1-6 days", 
                           "7-30 days", "31+ days")
  sim2 <- reshape2::melt(sim, id.vars = c(1,2), value.name="Frequency")
  sim2$variable <- factor(sim2$variable, levels = c("Diagnosed_No","Diagnosed_Yes"))
  p <- ggplot(sim2, aes(x = Exposure, y = Frequency, color = variable, fill = variable)) + 
    geom_bar(stat="identity")+facet_grid(.~Genotype) + theme_bw()
  ggplotly(p)
})

```


## 4. Conclusion
